design/keycloak.x/operator.md

```diff
@@ -0,0 +1,128 @@
+
+#### Static configuration through Kubernetes API
+
+The operator will leverage dedicated CRD(s), initially, there will be only one `Realm` CRD directly translated from Keycloak's [RealmRepresentation](https://github.com/keycloak/keycloak/blob/c7134fd5390d7c650b3dfd4bd2a2855157042271/core/src/main/java/org/keycloak/representations/idm/RealmRepresentation.java). A Realm includes all subresources. As a result, it is going to be possible to configure every object in Keycloak through this CR even though for some of them it won't be recommended (e.g. Users). To implement this, the operator will simply translate the CRs to YAML files and mount them to Keycloak pods, again leveraging Store.X.
+
+It will be possible to store any credentials as Secrets in K8s leveraging [Keycloak Vault functionality](https://www.keycloak.org/docs/latest/server_admin/index.html#_vault-administration) where possible.
+
+It's purpose of the upcoming Store.X initiative to provide a full-fledged static configuration backend for Keycloak but there will be a mid-term preview to enable bulk imports at startup time leveraging the REST api.
+
+### Keycloak versions alignment
+
+The operator will have its release cycle aligned with Keycloak. Each operator version will support only one corresponding Keycloak version.
+
+### Upgrades
+
+In order to upgrade to a newer Keycloak version, the operator will be upgraded first to ensure full compatibility with the operand.
+
+If custom Keycloak image is not used, the operator will use a default base image. After the operator is upgraded, it automatically upgrades Keycloak too using a newer base image.
+
+In case a custom Keycloak image is used, the image will need to be rebuilt to perform the upgrade. This is not going to be operator's responsibility as building a custom image often requires a complex process incl. CI/CD pipelines. After the operator is upgraded, it won't manage any existing Keycloak instances until its custom image is manually rebuilt using the right Keycloak base image aligned with the operator and updated and the image coordinates are updated in the CR.
+
+Store.X will allow zero-downtime rolling upgrades (a Keycloak upgrade performed pod by pod) that will ensure that Keycloak cluster will remain available even when upgrade fails on one of the pods. 
+
+### Reaugmentation process in Kubernetes
+
+We will be leveraging Kubernetes volumes to act as "caches" for the [augmented/configured](https://quarkus.io/guides/reaugmentation) version of Keycloak.
+An initial POC to show the concept has been drafted here:
+https://github.com/andreaTP/poc-mutable-jar
+
+We will use Kubernetes volumes to cache the augmented version of the binaries.
+The artifacts in the kubernetes volume will be produced by an init-container and the operation is going to result in a noop in case the volume has already been populated by a compatible augmentation.
```

Comment:
What happens when the reaugmentation fails for some reason? retry mechanism? messages, etc?

Reply:
I can imagine a "reaugmentation fault" is the same as "pod won't start", i.e. readiness probes will be failing.